packages/lesswrong/server/notificationCallbacksHelpers.tsx

```diff
@@ -0,0 +1,261 @@
+import Notifications from '../lib/collections/notifications/collection';
+import Messages from '../lib/collections/messages/collection';
+import { messageGetLink } from '../lib/helpers';
+import Subscriptions from '../lib/collections/subscriptions/collection';
+import Users from '../lib/collections/users/collection';
+import { userGetProfileUrl } from '../lib/collections/users/helpers';
+import { Posts } from '../lib/collections/posts';
+import { postGetPageUrl } from '../lib/collections/posts/helpers';
+import { Comments } from '../lib/collections/comments/collection'
+import { commentGetPageUrlFromDB } from '../lib/collections/comments/helpers'
+import './emailComponents/EmailWrapper';
+import './emailComponents/NewPostEmail';
+import './emailComponents/PostNominatedEmail';
+import './emailComponents/PrivateMessagesEmail';
+import { DebouncerTiming } from './debouncer';
+import { ensureIndex } from '../lib/collectionUtils';
+import { getNotificationTypeByName } from '../lib/notificationTypes';
+import { notificationDebouncers } from './notificationBatching';
+import { defaultNotificationTypeSettings } from '../lib/collections/users/custom_fields';
+import * as _ from 'underscore';
+import { createMutator } from './vulcan-lib/mutators';
+import keyBy from 'lodash/keyBy';
+import TagRels from '../lib/collections/tagRels/collection';
+
+/**
+ * Return a list of users (as complete user objects) subscribed to a given
+ * document. This is the union of users who have subscribed to it explicitly,
+ * and users who were subscribed to it by default and didn't suppress the
+ * subscription.
+ *
+ * documentId: The document to look for subscriptions to.
+ * collectionName: The collection the document to look for subscriptions to is in.
+ * type: The type of subscription to check for.
+ * potentiallyDefaultSubscribedUserIds: (Optional) An array of user IDs for
+ *   users who are potentially subscribed to this document by default, eg
+ *   because they wrote the post being replied to or are an organizer of the
+ *   group posted in.
+ * userIsDefaultSubscribed: (Optional. User=>bool) If
+ *   potentiallyDefaultSubscribedUserIds is given, takes a user and returns
+ *   whether they would be default-subscribed to this document.
+ */
+ export async function getSubscribedUsers({
+  documentId, collectionName, type,
+  potentiallyDefaultSubscribedUserIds=null, userIsDefaultSubscribed=null
+}: {
+  documentId: string|null,
+  collectionName: CollectionNameString,
+  type: string,
+  potentiallyDefaultSubscribedUserIds?: null|Array<string>,
+  userIsDefaultSubscribed?: null|((u:DbUser)=>boolean),
+}) {
+  if (!documentId) {
+    return [];
+  }
+  
+  const subscriptions = await Subscriptions.find({documentId, type, collectionName, deleted: false, state: 'subscribed'}).fetch()
+  const explicitlySubscribedUserIds = _.pluck(subscriptions, 'userId')
```

Comment:
I think you should be able to just have Mongo only return the user IDs: https://docs.mongodb.com/manual/tutorial/project-fields-from-query-results/

(Edit: I see you didn't actually write this code, feel free to leave it if you want.)